File: tetris.py
import pygame
import random
import sys

pygame.init()

BOARD_WIDTH = 10
BOARD_HEIGHT = 20
BLOCK_SIZE = 30
WINDOW_WIDTH = BOARD_WIDTH * BLOCK_SIZE + 200
WINDOW_HEIGHT = BOARD_HEIGHT * BLOCK_SIZE + 100

BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
YELLOW = (255, 255, 0)
MAGENTA = (255, 0, 255)
CYAN = (0, 255, 255)
ORANGE = (255, 165, 0)
GRAY = (128, 128, 128)

COLORS = [
    BLACK,    # 0: empty
    RED,      # 1: I piece
    GREEN,    # 2: O piece
    BLUE,     # 3: T piece
    YELLOW,   # 4: S piece
    MAGENTA,  # 5: Z piece
    CYAN,     # 6: J piece
    ORANGE    # 7: L piece
]

PIECES = [
    [],  # 0: empty
    [    # 1: I piece
        [0, 0, 0, 0],
        [1, 1, 1, 1],
        [0, 0, 0, 0],
        [0, 0, 0, 0]
    ],
    [    # 2: O piece
        [2, 2],
        [2, 2]
    ],
    [    # 3: T piece
        [0, 3, 0],
        [3, 3, 3],
        [0, 0, 0]
    ],
    [    # 4: S piece
        [0, 4, 4],
        [4, 4, 0],
        [0, 0, 0]
    ],
    [    # 5: Z piece
        [5, 5, 0],
        [0, 5, 5],
        [0, 0, 0]
    ],
    [    # 6: J piece
        [6, 0, 0],
        [6, 6, 6],
        [0, 0, 0]
    ],
    [    # 7: L piece
        [0, 0, 7],
        [7, 7, 7],
        [0, 0, 0]
    ]
]


class Tetris:
    def __init__(self):
        self.screen = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))
        pygame.display.set_caption('テトリス')
        self.clock = pygame.time.Clock()
        self.font = pygame.font.Font(None, 36)
        
        self.board = [[0] * BOARD_WIDTH for _ in range(BOARD_HEIGHT)]
        self.score = 0
        self.level = 1
        self.lines = 0
        self.drop_time = 0
        self.drop_interval = 1000
        
        self.current_piece = self.create_piece()
        self.next_piece = self.create_piece()
        
    def create_piece(self):
        piece_type = random.randint(1, 7)
        return {
            'type': piece_type,
            'shape': [row[:] for row in PIECES[piece_type]],
            'x': BOARD_WIDTH // 2 - len(PIECES[piece_type][0]) // 2,
            'y': 0
        }
    
    def is_valid_position(self, shape, x, y):
        for row in range(len(shape)):
            for col in range(len(shape[row])):
                if shape[row][col] != 0:
                    board_x = x + col
                    board_y = y + row
                    
                    if (board_x < 0 or board_x >= BOARD_WIDTH or 
                        board_y >= BOARD_HEIGHT or
                        (board_y >= 0 and self.board[board_y][board_x] != 0)):
                        return False
        return True
    
    def rotate_piece(self, shape):
        rows = len(shape)
        cols = len(shape[0])
        rotated = [[0] * rows for _ in range(cols)]
        
        for i in range(rows):
            for j in range(cols):
                rotated[j][rows - 1 - i] = shape[i][j]
        
        return rotated
    
    def move_piece(self, dx, dy):
        new_x = self.current_piece['x'] + dx
        new_y = self.current_piece['y'] + dy
        
        if self.is_valid_position(self.current_piece['shape'], new_x, new_y):
            self.current_piece['x'] = new_x
            self.current_piece['y'] = new_y
            return True
        elif dy > 0:
            self.place_piece()
            return False
        return False
    
    def rotate_current_piece(self):
        rotated = self.rotate_piece(self.current_piece['shape'])
        if self.is_valid_position(rotated, self.current_piece['x'], self.current_piece['y']):
            self.current_piece['shape'] = rotated
    
    def hard_drop(self):
        while self.move_piece(0, 1):
            pass
    
    def place_piece(self):
        for row in range(len(self.current_piece['shape'])):
            for col in range(len(self.current_piece['shape'][row])):
                if self.current_piece['shape'][row][col] != 0:
                    board_y = self.current_piece['y'] + row
                    board_x = self.current_piece['x'] + col
                    
                    if board_y >= 0:
                        self.board[board_y][board_x] = self.current_piece['type']
        
        self.clear_lines()
        self.current_piece = self.next_piece
        self.next_piece = self.create_piece()
        
        if not self.is_valid_position(self.current_piece['shape'], 
                                    self.current_piece['x'], 
                                    self.current_piece['y']):
            self.game_over()
    
    def clear_lines(self):
        lines_cleared = 0
        row = BOARD_HEIGHT - 1
        
        while row >= 0:
            if all(cell != 0 for cell in self.board[row]):
                del self.board[row]
                self.board.insert(0, [0] * BOARD_WIDTH)
                lines_cleared += 1
            else:
                row -= 1
        
        if lines_cleared > 0:
            self.lines += lines_cleared
            self.score += lines_cleared * 100 * self.level
            self.level = self.lines // 10 + 1
            self.drop_interval = max(50, 1000 - (self.level - 1) * 50)
    
    def game_over(self):
        print(f"ゲームオーバー！スコア: {self.score}")
        self.board = [[0] * BOARD_WIDTH for _ in range(BOARD_HEIGHT)]
        self.score = 0
        self.level = 1
        self.lines = 0
        self.drop_interval = 1000
        self.current_piece = self.create_piece()
        self.next_piece = self.create_piece()
    
    def handle_events(self):
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return False
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT:
                    self.move_piece(-1, 0)
                elif event.key == pygame.K_RIGHT:
                    self.move_piece(1, 0)
                elif event.key == pygame.K_DOWN:
                    self.move_piece(0, 1)
                elif event.key == pygame.K_UP:
                    self.rotate_current_piece()
                elif event.key == pygame.K_SPACE:
                    self.hard_drop()
        return True
    
    def update(self, dt):
        self.drop_time += dt
        
        if self.drop_time >= self.drop_interval:
            self.move_piece(0, 1)
            self.drop_time = 0
    
    def draw_board(self):
        for row in range(BOARD_HEIGHT):
            for col in range(BOARD_WIDTH):
                x = col * BLOCK_SIZE
                y = row * BLOCK_SIZE
                
                if self.board[row][col] != 0:
                    color = COLORS[self.board[row][col]]
                    pygame.draw.rect(self.screen, color, (x, y, BLOCK_SIZE, BLOCK_SIZE))
                
                pygame.draw.rect(self.screen, GRAY, (x, y, BLOCK_SIZE, BLOCK_SIZE), 1)
    
    def draw_current_piece(self):
        if not self.current_piece:
            return
        
        color = COLORS[self.current_piece['type']]
        
        for row in range(len(self.current_piece['shape'])):
            for col in range(len(self.current_piece['shape'][row])):
                if self.current_piece['shape'][row][col] != 0:
                    x = (self.current_piece['x'] + col) * BLOCK_SIZE
                    y = (self.current_piece['y'] + row) * BLOCK_SIZE
                    
                    pygame.draw.rect(self.screen, color, (x, y, BLOCK_SIZE, BLOCK_SIZE))
                    pygame.draw.rect(self.screen, GRAY, (x, y, BLOCK_SIZE, BLOCK_SIZE), 1)
    
    def draw_info(self):
        info_x = BOARD_WIDTH * BLOCK_SIZE + 10
        
        score_text = self.font.render(f"スコア: {self.score}", True, WHITE)
        self.screen.blit(score_text, (info_x, 50))
        
        level_text = self.font.render(f"レベル: {self.level}", True, WHITE)
        self.screen.blit(level_text, (info_x, 90))
        
        lines_text = self.font.render(f"ライン: {self.lines}", True, WHITE)
        self.screen.blit(lines_text, (info_x, 130))
        
        controls_text = [
            "操作方法:",
            "← →: 移動",
            "↓: 落下速度アップ",
            "↑: 回転",
            "スペース: ハードドロップ"
        ]
        
        for i, text in enumerate(controls_text):
            font_size = 24 if i == 0 else 18
            font = pygame.font.Font(None, font_size)
            control_text = font.render(text, True, WHITE)
            self.screen.blit(control_text, (info_x, 200 + i * 25))
    
    def draw(self):
        self.screen.fill(BLACK)
        self.draw_board()
        self.draw_current_piece()
        self.draw_info()
        pygame.display.flip()
    
    def run(self):
        running = True
        
        while running:
            dt = self.clock.tick(60)
            
            running = self.handle_events()
            self.update(dt)
            self.draw()
        
        pygame.quit()
        sys.exit()


if __name__ == "__main__":
    game = Tetris()
    game.run()

File: tetris.js
class Tetris {
    constructor() {
        this.canvas = document.getElementById('tetris');
        this.ctx = this.canvas.getContext('2d');
        this.scoreElement = document.getElementById('score');
        this.levelElement = document.getElementById('level');
        this.linesElement = document.getElementById('lines');
        
        this.BOARD_WIDTH = 10;
        this.BOARD_HEIGHT = 20;
        this.BLOCK_SIZE = 30;
        
        this.board = this.createBoard();
        this.score = 0;
        this.level = 1;
        this.lines = 0;
        this.dropTime = 0;
        this.dropInterval = 1000;
        
        this.currentPiece = null;
        this.nextPiece = null;
        
        this.colors = [
            '#000000', // 0: empty
            '#FF0000', // 1: I piece
            '#00FF00', // 2: O piece
            '#0000FF', // 3: T piece
            '#FFFF00', // 4: S piece
            '#FF00FF', // 5: Z piece
            '#00FFFF', // 6: J piece
            '#FFA500'  // 7: L piece
        ];
        
        this.pieces = [
            [], // 0: empty
            [   // 1: I piece
                [0,0,0,0],
                [1,1,1,1],
                [0,0,0,0],
                [0,0,0,0]
            ],
            [   // 2: O piece
                [2,2],
                [2,2]
            ],
            [   // 3: T piece
                [0,3,0],
                [3,3,3],
                [0,0,0]
            ],
            [   // 4: S piece
                [0,4,4],
                [4,4,0],
                [0,0,0]
            ],
            [   // 5: Z piece
                [5,5,0],
                [0,5,5],
                [0,0,0]
            ],
            [   // 6: J piece
                [6,0,0],
                [6,6,6],
                [0,0,0]
            ],
            [   // 7: L piece
                [0,0,7],
                [7,7,7],
                [0,0,0]
            ]
        ];
        
        this.initGame();
    }
    
    createBoard() {
        return Array(this.BOARD_HEIGHT).fill(0).map(() => Array(this.BOARD_WIDTH).fill(0));
    }
    
    createPiece() {
        const pieceType = Math.floor(Math.random() * 7) + 1;
        return {
            type: pieceType,
            shape: this.pieces[pieceType],
            x: Math.floor(this.BOARD_WIDTH / 2) - Math.floor(this.pieces[pieceType][0].length / 2),
            y: 0
        };
    }
    
    initGame() {
        this.currentPiece = this.createPiece();
        this.nextPiece = this.createPiece();
        this.setupControls();
        this.gameLoop();
    }
    
    setupControls() {
        document.addEventListener('keydown', (e) => {
            switch(e.code) {
                case 'ArrowLeft':
                    this.movePiece(-1, 0);
                    break;
                case 'ArrowRight':
                    this.movePiece(1, 0);
                    break;
                case 'ArrowDown':
                    this.movePiece(0, 1);
                    break;
                case 'ArrowUp':
                    this.rotatePiece();
                    break;
                case 'Space':
                    this.hardDrop();
                    break;
            }
        });
    }
    
    movePiece(dx, dy) {
        if (!this.currentPiece) return;
        
        const newX = this.currentPiece.x + dx;
        const newY = this.currentPiece.y + dy;
        
        if (this.isValidPosition(this.currentPiece.shape, newX, newY)) {
            this.currentPiece.x = newX;
            this.currentPiece.y = newY;
        } else if (dy > 0) {
            this.placePiece();
        }
    }
    
    rotatePiece() {
        if (!this.currentPiece) return;
        
        const rotated = this.rotateMatrix(this.currentPiece.shape);
        if (this.isValidPosition(rotated, this.currentPiece.x, this.currentPiece.y)) {
            this.currentPiece.shape = rotated;
        }
    }
    
    rotateMatrix(matrix) {
        const rows = matrix.length;
        const cols = matrix[0].length;
        const rotated = Array(cols).fill(0).map(() => Array(rows).fill(0));
        
        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                rotated[j][rows - 1 - i] = matrix[i][j];
            }
        }
        
        return rotated;
    }
    
    hardDrop() {
        if (!this.currentPiece) return;
        
        while (this.isValidPosition(this.currentPiece.shape, this.currentPiece.x, this.currentPiece.y + 1)) {
            this.currentPiece.y++;
        }
        this.placePiece();
    }
    
    isValidPosition(shape, x, y) {
        for (let row = 0; row < shape.length; row++) {
            for (let col = 0; col < shape[row].length; col++) {
                if (shape[row][col] !== 0) {
                    const boardX = x + col;
                    const boardY = y + row;
                    
                    if (boardX < 0 || boardX >= this.BOARD_WIDTH || 
                        boardY >= this.BOARD_HEIGHT ||
                        (boardY >= 0 && this.board[boardY][boardX] !== 0)) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    
    placePiece() {
        if (!this.currentPiece) return;
        
        for (let row = 0; row < this.currentPiece.shape.length; row++) {
            for (let col = 0; col < this.currentPiece.shape[row].length; col++) {
                if (this.currentPiece.shape[row][col] !== 0) {
                    const boardY = this.currentPiece.y + row;
                    const boardX = this.currentPiece.x + col;
                    
                    if (boardY >= 0) {
                        this.board[boardY][boardX] = this.currentPiece.type;
                    }
                }
            }
        }
        
        this.clearLines();
        this.currentPiece = this.nextPiece;
        this.nextPiece = this.createPiece();
        
        if (!this.isValidPosition(this.currentPiece.shape, this.currentPiece.x, this.currentPiece.y)) {
            this.gameOver();
        }
    }
    
    clearLines() {
        let linesCleared = 0;
        
        for (let row = this.BOARD_HEIGHT - 1; row >= 0; row--) {
            if (this.board[row].every(cell => cell !== 0)) {
                this.board.splice(row, 1);
                this.board.unshift(Array(this.BOARD_WIDTH).fill(0));
                linesCleared++;
                row++;
            }
        }
        
        if (linesCleared > 0) {
            this.lines += linesCleared;
            this.score += linesCleared * 100 * this.level;
            this.level = Math.floor(this.lines / 10) + 1;
            this.dropInterval = Math.max(50, 1000 - (this.level - 1) * 50);
            
            this.updateDisplay();
        }
    }
    
    updateDisplay() {
        this.scoreElement.textContent = this.score;
        this.levelElement.textContent = this.level;
        this.linesElement.textContent = this.lines;
    }
    
    gameOver() {
        alert('ゲームオーバー！\nスコア: ' + this.score);
        this.board = this.createBoard();
        this.score = 0;
        this.level = 1;
        this.lines = 0;
        this.dropInterval = 1000;
        this.updateDisplay();
        this.initGame();
    }
    
    draw() {
        this.ctx.fillStyle = '#111';
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        
        this.drawBoard();
        this.drawCurrentPiece();
    }
    
    drawBoard() {
        for (let row = 0; row < this.BOARD_HEIGHT; row++) {
            for (let col = 0; col < this.BOARD_WIDTH; col++) {
                if (this.board[row][col] !== 0) {
                    this.ctx.fillStyle = this.colors[this.board[row][col]];
                    this.ctx.fillRect(
                        col * this.BLOCK_SIZE,
                        row * this.BLOCK_SIZE,
                        this.BLOCK_SIZE,
                        this.BLOCK_SIZE
                    );
                    
                    this.ctx.strokeStyle = '#333';
                    this.ctx.lineWidth = 1;
                    this.ctx.strokeRect(
                        col * this.BLOCK_SIZE,
                        row * this.BLOCK_SIZE,
                        this.BLOCK_SIZE,
                        this.BLOCK_SIZE
                    );
                }
            }
        }
    }
    
    drawCurrentPiece() {
        if (!this.currentPiece) return;
        
        this.ctx.fillStyle = this.colors[this.currentPiece.type];
        
        for (let row = 0; row < this.currentPiece.shape.length; row++) {
            for (let col = 0; col < this.currentPiece.shape[row].length; col++) {
                if (this.currentPiece.shape[row][col] !== 0) {
                    const x = (this.currentPiece.x + col) * this.BLOCK_SIZE;
                    const y = (this.currentPiece.y + row) * this.BLOCK_SIZE;
                    
                    this.ctx.fillRect(x, y, this.BLOCK_SIZE, this.BLOCK_SIZE);
                    
                    this.ctx.strokeStyle = '#333';
                    this.ctx.lineWidth = 1;
                    this.ctx.strokeRect(x, y, this.BLOCK_SIZE, this.BLOCK_SIZE);
                }
            }
        }
    }
    
    update(deltaTime) {
        this.dropTime += deltaTime;
        
        if (this.dropTime >= this.dropInterval) {
            this.movePiece(0, 1);
            this.dropTime = 0;
        }
    }
    
    gameLoop(currentTime = 0) {
        const deltaTime = currentTime - (this.lastTime || 0);
        this.lastTime = currentTime;
        
        this.update(deltaTime);
        this.draw();
        
        requestAnimationFrame((time) => this.gameLoop(time));
    }
}

new Tetris();